### ISSUE TITLE
Mod Layer Collision

### ISSUE THREAD

**Issue by KiwiHawk** [modded-factorio/LandfillPainting#13]
Various issues probably caused by layer collision checks. With AAI and / Krastorio 2 / Space Exploration:

https://mods.factorio.com/mod/LandfillPainting/discussion/6293a7c820c9f2d7bfe18bc6
> Im using a bunch of mods including rampant, alien biomes and krastorio2. I cannot drive vehicles over stone bricks

https://mods.factorio.com/mod/LandfillPainting/discussion/626b93cbccd725edeac52f4e
> Was playing a Krastorio run with AAI vehicles included, and after enabling this mod I stumbled into two bugs, one, the 'cant place pipes in artificial ground' someone else has reported too.
> 
> and two, AAI vehicles with AI enabled refuse to path over any artificial ground, even the creep on biter nests that Krastorio adds.
> 
> Disabling this mod solved both issues.

https://mods.factorio.com/mod/LandfillPainting/discussion/617539693a29d159a1b97f3a
> I dont know if this is really a bug, but when Landfill Painting is active I cant place any pipes on stone bricks, concrete or refined concrete. If this is intended: Why? If not: Please fix it because this mod is awesome.

https://mods.factorio.com/mod/LandfillPainting/discussion/5ecb8d34b7d53d000bade69a
> Space exploration uses layers 14 and 15 to facilitate space\spaceship collision. Your mod adding layer 15 to all player placed tiles breaks it.
> If you could add an exception for tiles that start with 'se-', it should fix it.

https://mods.factorio.com/mod/LandfillPainting/discussion/62bcffb48ed1b6ac27b67aae
Hello, I have been playing through the seablock modpack and the game will not let me place piping on any type of concrete. I turned off this mod and was able to place the piping on the concrete but in doing that I lose all my landfill.
[New Save.zip](https://github.com/KiwiHawk/LandfillPainting/files/9051839/New.Save.zip)

Also, placing blasting charges on existing islands in Sea Block is free. Doesn't consume the blasting charge. Not sure if this applies to other waterfill mods too.

**Comment by Eurymachus** [comment]
Began a Sea Block run through last week, built a car and can't drive the car over paths, can't build the car on the path/concrete and when building stone/concrete under a car the car goes Poof!

Managed to dig through and find this issue logged from the mod discussions, disabled the mod and the issue went away.

Of course if I leave the mod disabled I cannot effectively craft landfill lol.

If you need any more info, I'd be happy to provide :)

**Comment by KiwiHawk** [comment]
> Began a Sea Block run through last week, built a car and can't drive the car over paths, can't build the car on the path/concrete and when building stone/concrete under a car the car goes Poof!
> 
> Managed to dig through and find this issue logged from the mod discussions, disabled the mod and the issue went away.
> 
> Of course if I leave the mod disabled I cannot effectively craft landfill lol.
> 
> If you need any more info, I'd be happy to provide :)

Which other mods do you have enabled, other than the Sea Block Pack? With just the Sea Block Pack mods it works fine so there's definitely some other mod affecting things. My guess is Transport Drones? 🤔

**Comment by Eurymachus** [comment]
Your point made me think a little, so I went through my list.

TL;DR Disabling the 'Combat Mechanics Overhaul' mod fixed the issue so I'm going to dig deeper into it's config.

Long Story;
I don't have transport drones but, I figured I'd take a look at certain mods and see if there was something that could possibly affect collision.

Combat Mechanics Overhaul, literally mentions the word 'collision' in the description and how it alters items/units/vehicles etc to have specific tags so they can be highly configurable to create individual interactions between diffferent game elememts.

Running that mod without Landfill and everything is fine.
Disabling that mod and running with Landfill and everything is fine.
Both together is no bueno.

I'm going to dig into the configurations and see if there is something for interactions between vehicles, paving and different land types.

**Comment by Eurymachus** [comment]
> Your point made me think a little, so I went through my list.
> 
> TL;DR Disabling the 'Combat Mechanics Overhaul' mod fixed the issue so I'm going to dig deeper into it's config.
> 
> Long Story; I don't have transport drones but, I figured I'd take a look at certain mods and see if there was something that could possibly affect collision.
> 
> Combat Mechanics Overhaul, literally mentions the word 'collision' in the description and how it alters items/units/vehicles etc to have specific tags so they can be highly configurable to create individual interactions between different game elements.
> 
> Running that mod without Landfill and everything is fine. Disabling that mod and running with Landfill and everything is fine. Both together is no bueno.
> 
> I'm going to dig into the configurations and see if there is something for interactions between vehicles, paving and different land types.

Config is very basic. There doesn't appear to be any open source for CMO so I'm removing it completely. It's largely useless any way for a SeaBlock run through, it's just left from my AngelBob pack that I put together previously :D

Still, interesting that they don't play nice together.

**Comment by KiwiHawk** [comment]
I'm a bit short of time currently. But this is on my list of things to investigate. Hopefully it should be as simple as changing which layer we use.

As a quick fix, I could mark Combat Mechanics Overhaul as incompatible. I'd rather not do that if I can avoid it though.

**Comment by heinwessels** [comment]
I saw this problem being reported on the SE Discord. Best is to choose a collision layer dynamically using some util. For example this tool provided by Wube.

https://github.com/wube/factorio-data/blob/20a5999124a021dd17045e770c8b1e434304c68e/core/lualib/collision-mask-util.lua#L278C16-L284
